### ISSUE TITLE
create an msys2 package

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1956]
msys2 looks like good stuff. we ought get ourselves added to their repository as a package.

https://packages.msys2.org/queue

**Comment by WSLUser** [comment]
Actually you want Cygwin first. They're almost the same but there's some differences and Cygwin often acts as the upstream for Msys2 these days (Cygwin was originally derived from Msys2 if I'm recalling correctly). Mintty is a Cygwin package though is also available on Msys2, which would be the terminal you're targeting for that application suite. )

Also don't allow building for this stop you from also building a native Windows package.

**Comment by michaelsbradleyjr** [comment]
> we ought get ourselves added to their repository as a package

I think that would be great. Typically, packages have a build for each of the [MSYS2 environments](https://www.msys2.org/docs/environments/), though I'm guessing some packages don't have builds for all of them.

Maybe the first step would be landing a package for UCRT64 or MINGW64, and then later on see if it's possible to get builds working for the others.

---
@WSLUser 
> Actually you want Cygwin first.

Support for building in Cygwin and creating a package for Cygwin would be fine. But I don't think its being done first is particularly desirable or important.

> Mintty is a Cygwin package though is also available on Msys2, which would be the terminal you're targeting for that application suite.

Software built with MSYS2 and notcurses would hopefully run in Windows Terminal and ConEMU, in addition to mintty, and others as well, e.g. Alacritty or Contour built for Windows.

> Cygwin was originally derived from Msys2

This is backwards; you may be thinking of MSYS instead of MSYS2. See below.

---

From [msys2.org](https://www.msys2.org/) — [*How does MSYS2 differ from Cygwin*](https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/):

In all cases, emphasis is mine.

> Cygwin and MSYS2 -- as projects -- have **significantly different goals**.

> Cygwin tries to bring a POSIX-compatible environment to Windows... MSYS2 tries to provide an environment for building **native Windows software**.

> [MSYS2's] shell and core tools exist mainly to allow **porting Unix programs to run natively on Windows (i.e. without requiring a POSIX emulation layer)**. MSYS2 doesn't try to duplicate Cygwin's efforts more than necessary, so the number of provided POSIX-emulated software is very small.

> **MSYS2 uses Pacman (known from Arch Linux) to manage its packages**... Cygwin comes only with Cygwin-dependent software. It uses its own package management system, commonly called setup.exe.

> MSYS2 releases might be **ahead of or behind** Cygwin releases.

From elsewhere on [msys2.org](https://www.msys2.org/):

> MSYS2 vs Cygwin
> The unixy tools in MSYS2 are directly based on Cygwin, so there is some overlap there. While Cygwin focuses on building Unix software on Windows as is, **MSYS2 focuses on building native software built against the Windows APIs**.

And there's more info related to understanding MSYS2 in comparison to Cygwin, et al. sprinkled around the other pages, e.g. [History](https://www.msys2.org/wiki/History/):

> **MSYS2 is an independent rewrite** of MSYS, based on modern Cygwin and MinGW-w64 with the aim of better **interoperability with native Windows software**.

---

This comment and my comments in #1953 will be my last comments attempting to clear up misinformation re: MSYS2. Going forward, I'll just link to my comments in this PR and the forthcoming comment on #1956.

**Comment by WSLUser** [comment]
Thanks for that refresher. Yeah I was thinking of Msys vs Msys2. I also knew there was overlap though.  Building for one should allow for building for the other though. I'm just not sure which one would be more likely to accept a package first. I would of though Cygwin because this project was originally created for Linux and that's Cygwin's thing as stated in those FAQS you quoted. I forgot Msys2 was intending to aim for native Windows software but I'll stand by what I said about the MS build chain being the proper official way for native Windows applications.

**Comment by dankamongmen** [comment]
#1975 would be a dependency of this; i'm uncomfortable submitting anything without unit testing behind it

**Comment by dankamongmen** [comment]
we're now passing `ctest` in our msys2 builds, so i guess it's time? i've had a really difficult time trying to get `notcurses-tester` built on windows, though.

**Comment by dankamongmen** [comment]
I think we're ready to move forward with this. We won't be building or running `notcurses-tester` just yet, but we can build the PoCs and run `ctest`. I think we don't want to install most of the binaries, actually, since almost all of the deliverable ones require MS dlls and thus don't run within MSYS2 proper (they require running in the external Windows install). We'd just be packaging up libnotcurses.

@michaelsbradleyjr , would you have any interest in doing this? I'm a bit ragged at the moment -- been working hard, and Pa Dank died a few days ago. I already did one for doctest, as you can see in #1975, which ought serve as a usable template. If you don't have time/inclination, I can go ahead and get it done.

**Comment by michaelsbradleyjr** [comment]
I'll review this weekend and let you know if I think I can / can't handle it, will certainly try my best to take care of it.

I'm sorry to hear of the passing of Pa Dank. I think I understand to whom that refers, and I offer my deepest condolences: *Requiem aeternam dona ei, Domine, et lux perpetua luceat ei. Requiescat in pace. Amen.*

**Comment by michaelsbradleyjr** [comment]
First step today, I caught up with `master` branch on my Windows 10 laptop and did some builds.

In the GHA workflow for Windows, there's still `-DUSE_DOCTEST=off`, though as you noted above you've already submitted a package for doctest and I've since installed it. However, if I don't pass `-DUSE_DOCTEST=off` to `cmake` then `make` fails:

![image](https://user-images.githubusercontent.com/194260/129447539-f3dc3f27-b5d6-4620-a642-c1a3033665ec.png)

That may be expected, but I wanted to check with you.

Similarly, I find that if I don't pass `-DUSE_READLINE=off` to `cmake` then `make` fails. In your instructions for building on Windows you note that the build never uses it, but I wasn't sure if it's expected that it needs to be explicitly disabled.

For a successful build and `ctest` (100% tests passed), I do:

```
$ git clean -dfx && mkdir build && cd build
$ cmake -G "MSYS Makefiles" -DCMAKE_BUILD_TYPE=Release -DUSE_DOCTEST=off -DUSE_PANDOC=off -DUSE_READLINE=off ..
$ make
$ ctest
```
---

As an experiment, I tried a build using MSYS2's `C:\mingw64` toolchain instead of the `C:\ucrt64` toolchain. `cmake` runs successfully but `make` fails right away owing to (I think) differences in what the toolchains support re: the C sources. Anyway, I thought it was worth a quick look.

---

> We won't be building or running `notcurses-tester` just yet, but we can build the PoCs and run `ctest`... We'd just be packaging up libnotcurses.

That seems very reasonable.

>  I think we don't want to install most of the binaries, actually, since almost all of the deliverable ones require MS dlls and thus don't run within MSYS2 proper (they require running in the external Windows install).

This part I'm not sure I understand, i.e. what you mean by "external Windows install".

In a follow-up comment I'll share some screenshots of running `notcurses-info.exe` in several terminal emulators and a mix of invoking the exe from `cmd.exe`, PowerShell, and MSYS2 UCRT64 bash.

**Comment by dankamongmen** [comment]
correct, `notcurses-tester` does not yet build despite doctest being available. it'll take some time to get it there. let's not block on it.

**Comment by michaelsbradleyjr** [comment]
Here are the screenshots. I'm not sure if they're helpful at all, but it was fun to experiment and see the progress you've made!

I'll start working on the package now.

In the same mintty+bash that I used to do the build:

![image](https://user-images.githubusercontent.com/194260/129448497-5dee05d9-712e-45c1-99fc-c045e15fdaed.png)

In Windows Terminal + MSYS2 UCRT64 Bash:

![image](https://user-images.githubusercontent.com/194260/129448531-24cf97a8-54e8-4a54-aa53-29cd81f1e946.png)

In Windows Terminal + `cmd.exe` and `set PATH=C:\ucrt64\ucrt64\bin;%PATH%`:

![image](https://user-images.githubusercontent.com/194260/129448606-a27cfabc-c4ef-4c60-b415-2f453865871c.png)

In Windows Terminal + PowerShell and `$env:Path += "C:\ucrt64\ucrt64\bin"`:

![image](https://user-images.githubusercontent.com/194260/129448659-ee811d47-6c92-40c3-8f08-26833d848440.png)

In ConEmu + `cmd.exe` and `set PATH=C:\ucrt64\ucrt64\bin;%PATH%`:

![image](https://user-images.githubusercontent.com/194260/129448706-bc8fa3da-a956-4fd6-aaad-7b8f7741b42f.png)

In Alacritty + PowerShell and `$env:Path += "C:\ucrt64\ucrt64\bin"`:

![image](https://user-images.githubusercontent.com/194260/129448748-2d628a5a-bb91-400a-be5b-0700ccc68fb1.png)

**Comment by dankamongmen** [comment]
> I'm sorry to hear of the passing of Pa Dank. I think I understand to whom that refers, and I offer my deepest condolences: _[REQUIEM aeternam](https://www.preces-latinae.org/thesaurus/Varia/Requiem.html) dona ei, Domine, et lux perpetua luceat ei. Requiescat in pace. Amen._

thanks mang; he was a good one

**Comment by dankamongmen** [comment]
yeah there are definitely still some warnings, some of them quite real problems. the vast majority of the output is just one warning in `compat.c` being repeated dozens of times. i'll get to them.

**Comment by michaelsbradleyjr** [comment]
See: https://github.com/michaelsbradleyjr/MINGW-packages/blob/notcurses/mingw-w64-notcurses/PKGBUILD.

I haven't tested it locally yet, am about to do so. If it all works out I'll submit a PR.

There are some things I'm not certain about:
* Is it correct at this time to limit arch to `arch=('x86_64')`?
* Any idea if we should be limiting it somehow to only ucrt64?
* Is `pkgdesc` as you prefer it?
* Is the use of make var `DESTDIR` in `package()` something particular to doctest and so not appropriate for notcurses? Also, per what you expressed above, we just want to install the DLLs, right? If so, maybe `make install` isn't appropriate at all and I should just be copying the DLLs. I poked in the Makefile generated by cmake, but maybe there's a make target that would be appropriate for only installing the DLLs and I didn't notice it.

**Comment by michaelsbradleyjr** [comment]
Ah, I see DESTDIR is a cmake thing: https://cmake.org/cmake/help/latest/envvar/DESTDIR.html.

**Comment by dankamongmen** [comment]
> See: https://github.com/michaelsbradleyjr/MINGW-packages/blob/notcurses/mingw-w64-notcurses/PKGBUILD.
> I haven't tested it locally yet, am about to do so. If it all works out I'll submit a PR.

awesome!

> * Is it correct at this time to limit arch to `arch=('x86_64')`?

so far as i know, we ought work on all architectures. alpine, fedora, and debian all build us on at least four arches each, and run the unit test.

> * Any idea if we should be limiting it somehow to only ucrt64?

i'm not sure what you mean here. as opposed to a msys package? as opposed to a 32 bit? i know we're a UCRT (mingw64) and not msys package. as for 32 bit....i'd leave it off for now. i haven't verified any of it on any 32-bit machines, to the best of my knowledge.

> * Is `pkgdesc` as you prefer it?

looks good to me!

> * Is the use of make var `DESTDIR` in `package()` something particular to doctest and so not appropriate for notcurses? Also, per what you expressed above, we just want to install the DLLs, right? If so, maybe `make install` isn't appropriate at all and I should just be copying the DLLs. I poked in the Makefile generated by cmake, but maybe there's a make target that would be appropriate for only installing the DLLs and I didn't notice it.

so....`notcurses-info`, `notcurses-demo`, etc. don't work within the MSYS2 environment as far as i'm aware, as they require real Windows UCRT DLLs which MSYS can't distribute. that's why i figure they shouldn't be installed. if i'm wrong on that, they should. and maybe they should either way? you might want to bring the issue to the attention of the MSYS people and ask them. but yeah, what needs get installed are:

* DLLs
* header files

the little PoC binaries aren't installed on any platform, and shouldn't be, but they're not part of the install target.

really appreciate you getting this, homie!

**Comment by michaelsbradleyjr** [comment]
> i'm not sure what you mean here. as opposed to a msys package?

I meant distinguishing between the `UCRT64` and `MINGW64` [environments](https://www.msys2.org/docs/environments/).  At present, notcurses doesn't build for the latter, only the former.

**Comment by michaelsbradleyjr** [comment]
Okay it works now: https://github.com/michaelsbradleyjr/MINGW-packages/blob/notcurses/mingw-w64-notcurses/PKGBUILD.

BUT, there was a problem with `v2.3.13` such that to build, check, and install the MSYS2 package locally I needed to have PKGBUILD download an archive for the `master` branch instead. With `v2.3.13`, cmake has a problem finding some sources and the build fails; but there's no such problem with `master`. However, I don't think `master` is suitable for a PR to MINGW-packages repo.

So we'll hold off until the next release?

Also, worth noting, for unknown reasons the `check()` step froze and after a few minutes I had to ctrl+c. But on the next attempt it worked okay (I hadn't changed anything). Maybe it's a bug in the `makepkg-mingw` tool.

**Comment by dankamongmen** [comment]
yes the first possible mingw release would be 2.3.14. i just hoped to have the packaging slotted and ready to go by then, as it will likely be today or tomorrow following a particularly productive evening  https://github.com/dankamongmen/notcurses/discussions/1905.

https://www.youtube.com/watch?v=r2AHvAOKzkk

**Comment by michaelsbradleyjr** [comment]
I asked today in the MSYS2 discord server:

https://discord.com/channels/792780131906617355/794220101741838401/876529342438588496

![image](https://user-images.githubusercontent.com/194260/129489974-6a851894-9ab5-49ca-8311-ea78cf2dd940.png)

The response was:

![image](https://user-images.githubusercontent.com/194260/129489994-45ffbf54-b170-4b81-92f9-ffa843b47f20.png)

So... maybe when the PR is submitted a case can be made for a package that strictly works with `ucrt` but not `msvcrt `. But maybe not, in which case it would be important to ensure it can build in both the `UCRT64` and `MINGW64` environments.

Since the `CLANG64` environment also uses `ucrt` for the C library, then maybe notcurses will build in it without additional changes. Building in the MSYS2's `MSYS` environment is almost surely a non-goal, and that leaves only the `MINGW32` environment, which is also probably a non-goal for now.

**Comment by dankamongmen** [comment]
> So... maybe when the PR is submitted a case can be made for a package that strictly works with `ucrt` but not `msvcrt`. But maybe not, in which case it would be important to ensure notcurses can build in both the `UCRT64` and `MINGW64` environments.

hrmmm, does it? i'm not informed regarding the difference between the two. it would just be a difference between library support, right? in that case i would think we probably work on MINGW64, as it's presumably closer to POSIX than ucrt?

**Comment by michaelsbradleyjr** [comment]
> hrmmm, does it? i'm not informed regarding the difference between the two. it would just be a difference between library support, right? in that case i would think we probably work on MINGW64, as it's presumably closer to POSIX than ucrt?

The best thing is to give [MSYS2 environments](https://www.msys2.org/docs/environments/) a close read.

I can try to summarize:
* There is a base environment named `MSYS` that's comparable to cygwin and generally used to build and provide build-tooling like gcc.
* The other environments sit on top of that one and are used to build and provide native Windows software: `MINGW64`, `MINGW32`, `UCRT64 `, `CLANG64 `.
* The latter can differ from one another in terms of toolchain, architecture, C library, and C++ library.
  * `UCRT64`: gcc, ucrt, libstdc++
  * `MINGW64`: gcc, msvcrt, libstdc++

So for `MINGW64`, notcurses would need to work with msvcrt. The last section of the environments page has some bullet points re: the differences between msvcrt and ucrt.

I don't think `MINGW64` is necessarily closer to POSIX than `UCRT64`, but I could be wrong about that.

As mentioned above, I tried building notcurses in `MINGW64` and immediately met with errors re: the syntax of the sources, maybe related to:
> It isn't C99 compatible

I just noticed in the bullet points:
> It doesn't support the UTF-8 locale

So maybe that makes it a complete not-starter for notcurses, and could be a basis to justify to not building for that environment?

**Comment by dankamongmen** [comment]
the horror! the horror!

i can make it work, but i can't promise i can make it work well

**Comment by michaelsbradleyjr** [comment]
Is Notcurses not our ship of dreams and glorious hopes?

![image](https://user-images.githubusercontent.com/194260/129656279-9d57e033-fc00-485d-99c2-38b3613e56d1.png)

**Comment by michaelsbradleyjr** [comment]
I need to learn C, so I can take a bite..............

Best recommendation/s on paired-readings of Notcurses' sources by a C-novice (really a C-nincompoop), who also happens to be a fairly quick study? (where there's a will, there's a way, right?)

![image](https://user-images.githubusercontent.com/194260/129656485-f5f1a38d-11c3-4dc2-818b-1626f33a1b05.png)

**Comment by dankamongmen** [comment]
oh i mean i'm happy to do the C work, i just haven't tried building in that environment, and am going to be pressed for time the next two weeks more than i have been by ye olde dayjob. for all i know, it might build forthrightly. i guess what i'm thinking here is let's make a branch for this work (i doubt it'll be done for 2.3.14), and add a mingw64 workflow, and take a look at what all's broken so we know what kind of effort this is going to be.

having conferred with my MSFT homies this morning, the most important windows distribution target is WinGet in any case. if a mingw64 package is going to take eighty hours to do and involve further `#ifdef` mutilation of the codebase, i'm rapidly losing interest in it. in that case, i'd rather invest time in getting something into WinGet and possibly Chocolatey.

**Comment by michaelsbradleyjr** [comment]
> guess what i'm thinking here is let's make a branch for this work (i doubt it'll be done for 2.3.14), and add a mingw64 workflow, and take a look at what all's broken so we know what kind of effort this is going to be.

Sounds good. Also, when `2.3.14` is released, I think it's probably worth making a PR to MINGW-packages with an explanation that MINGW64 support may be possible in the future and that UCRT64 is a start. Since ucrt is a relatively recent thing for MSYS2, I suspect that as time goes by more package authors will be interested in skipping MINGW64/32 to avoid hoop-jumping to be compatible with both msvcrt and ucrt.

Before making the PR, I'll attempt to build in CLANG64, might just work and would be good to know either way.

> having conferred with my MSFT homies this morning, the most important windows distribution target is WinGet in any case. if a mingw64 package is going to take eighty hours to do and involve further #ifdef mutilation of the codebase, i'm rapidly losing interest in it. in that case, i'd rather invest time in getting something into WinGet and possibly Chocolatey.

Makes total sense. Hopefully the MSYS2 people will be open to non-MINGW64 packages. Regardless, availability via WinGet would be great!

I don't hear as much about Chocolatey these days, but another one I've used and been very pleased with is [Scoop](https://scoop.sh/). There are several [buckets](https://github.com/lukesampson/scoop/blob/master/buckets.json) available (each a GitHub repo) providing extensive collections of packages for Windows.
